### PR TITLE
Add byzantine integration tests

### LINKED ISSUE #3435: [Byzantine testing] - Add configurability for byzantine behaviour in integration tests
### What is this task and why do we need to work on it?

The byzantine behaviour framework should be integrated with our existing tests in a configurable way.

### What work will need to be done to complete this task?

_No response_

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

An integration test with some simple byzantine node behavior is implemented

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by ss-es**
Closes #3435 

### This PR: 
Integrates the byzantine trait into the existing integration test framework, configurable via the test metadata. In addition, `test_success` is replicated with nodes that double-vote and double-propose

### This PR does not: 
The twin handle implementation has some issues; this PR does not fix that. There are also some shutdown issues with byzantine nodes in `tokio`, so the test is gated for `async-std`.

### Key places to review: 
`testing/tests/tests_1/test_success.rs` for the extra double vote/propose test.

### CODE REVIEW

File: crates/testing/src/test_builder.rs [hunk 1]
```diff
@@ -78,6 +84,95 @@ pub struct TestDescription {
...
     pub solver: FakeSolverApiDescription,
+    /// nodes with byzantine behaviour
+    pub behaviour: Rc<dyn Fn(u64) -> Behaviour<TYPES, I>>,
+}
+
+#[derive(Debug)]
+pub enum Behaviour<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    Twins(&'static mut Box<dyn TwinsHandlerState<TYPES, I>>),
+    Single(&'static mut Box<dyn EventTransformerState<TYPES, I>>),
+    None,
```
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/testing/src/test_builder.rs:95]: Ok I guess this explains why we do `Box::leak(Box::new(Box::new()))` is there really something to be gained here with the static ref?  Why can't this type just be `mut Box<dyn ...>`?  Then we can get rid of the early leading of the boxed boxed.

**Review by bfish713 — COMMENTED** [review]
Looks good in general.  My main suggestion is to maybe rename all the byzantine stuff so it's clear in the code we are spawning something byzantine and to clean up a bit of the box stuff

File: crates/hotshot/src/tasks/mod.rs [hunk 2]
```diff
@@ -206,49 +210,87 @@ pub async fn add_consensus_tasks<TYPES: NodeType, I: NodeImplementation<TYPES>>(
...
+            hotshot: Arc::clone(&hotshot),
+            storage: Arc::clone(&hotshot.storage),
+            network: Arc::clone(&hotshot.network),
+            memberships: Arc::clone(&hotshot.memberships),
+        };
+
+        add_consensus_tasks::<TYPES, I>(&mut handle).await;
+        self.add_network_tasks(&mut handle).await;
+
+        handle
```
[the remark this answers is not in the document]
Comment (ss-es) [reply crates/hotshot/src/tasks/mod.rs:268]: Yeah, some parts were also part of e.g. the `SystemContext` impl but after refactoring I think it makes sense to maybe make a `byzantine` file under `hotshot/src/traits`.

I made an issue: https://github.com/EspressoSystems/HotShot/issues/3438

File: crates/testing/src/test_builder.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ss-es) [reply crates/testing/src/test_builder.rs]: Yeah, this was definitely not ideal. I originally tried `mut Box<dyn ...>` but there's a lifetime issue.

Basically, the way the byzantine behaviour sits in between network and consensus is by spawning two separate loops with the byzantine state, one for each direction (consensus -> network and network -> consensus). And there was just a cascade of problems to try to make the trait object-safe while having both of these loops get a copy of the state.

I can't require `Clone` from the state (requires `Self: Sized`) and I can't put it in an `RwLock::new()` (requires `Self: Sized`). The easiest thing was to just require the `'static` lifetime, but then everything needs to be `'static` and forces me to use `Box::leak` to create it. (And the second `Box::new()` is a mystery to me; the compiler was unhappy and told me to do it)

There's definitely a better way to do this, but I think I'd need to spend some time rewriting the loops stitching network <-> consensus.

[the remark this answers is not in the document]
Comment (ss-es) [reply crates/testing/src/test_builder.rs]: Renamed the `Behaviour` variants to `ByzantineTwins`, `Byzantine` and `Standard`!

[the remark this answers is not in the document]
Comment (ss-es) [reply crates/testing/src/test_builder.rs]: Update: removed the `'static` using your suggestion!

**Review by bfish713 — APPROVED** [review]

**Pull request merged**